File: common/src/main/java/dev/obscuria/fragmentum/core/v1/common/event/BaseEvent.java
package dev.obscuria.fragmentum.core.v1.common.event;

import com.google.common.collect.Lists;
import dev.obscuria.fragmentum.api.v1.common.event.Event;
import dev.obscuria.fragmentum.api.v1.common.event.EventHandler;
import dev.obscuria.fragmentum.api.v1.common.event.EventToken;
import org.jetbrains.annotations.ApiStatus;

import java.util.List;
import java.util.UUID;

@ApiStatus.Internal
public final class BaseEvent<T> implements Event<T>
{
    private final List<Registration<T>> registrations = Lists.newArrayList();
    private boolean dirty;

    @Override
    public EventToken register(T listener)
    {
        return register(1000, listener);
    }

    @Override
    public EventToken register(int priority, T listener)
    {
        final var token = new BaseEventToken(UUID.randomUUID(), this);
        this.registrations.add(new Registration<>(token, priority, listener));
        this.dirty = true;
        return token;
    }

    @Override
    public void unregister(EventToken token)
    {
        this.registrations.removeIf(registration -> registration.token.equals(token));
    }

    @Override
    public void broadcast(EventHandler<T> handler)
    {
        this.sortIfDirty();
        this.registrations.stream()
                .map(Registration::listener)
                .forEach(handler::handle);
    }

    private void sortIfDirty()
    {
        if (!dirty) return;
        this.dirty = false;
        this.registrations.sort(Registration::compareTo);
    }

    private record Registration<T>(EventToken token,
                                   int priority,
                                   T listener) implements Comparable<Registration<T>>
    {
        @Override
        public int compareTo(Registration<T> other)
        {
            return Integer.compare(this.priority, other.priority);
        }
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/core/tools/event/BaseEvent.kt
package dev.obscuria.fragmentum.core.tools.event

import com.google.common.collect.Lists
import dev.obscuria.fragmentum.api.tools.event.Event
import dev.obscuria.fragmentum.api.tools.event.EventHandler
import dev.obscuria.fragmentum.api.tools.event.EventToken
import java.util.*

class BaseEvent<T> : Event<T>
{
    private val registrations: MutableList<Registration<T>> = Lists.newArrayList()
    private var dirty = false

    override fun register(listener: T, priority: Int): EventToken
    {
        val token = BaseEventToken(UUID.randomUUID(), this)
        registrations.add(Registration(token, priority, listener))
        this.dirty = true
        return token
    }

    override fun unregister(token: EventToken)
    {
        registrations.removeIf { it.token == token }
    }

    override fun broadcast(handler: EventHandler<T>)
    {
        this.sortIfDirty()
        registrations.stream()
            .map { it.listener }
            .forEach(handler::handle)
    }

    private fun sortIfDirty()
    {
        if (!dirty) return
        registrations.sortWith(Registration<T>::compareTo)
        this.dirty = false
    }

    private class Registration<T>(
        val token: EventToken,
        val priority: Int,
        val listener: T
    ) : Comparable<Registration<T>>
    {
        override fun compareTo(other: Registration<T>): Int
        {
            return this.priority.compareTo(other.priority)
        }
    }
}